cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,468 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stdint.h>
+#include "cuda_utils.h"
+#include "utils.h"
+#include "vectorized.h"
+
+// This file is a shameless amalgamation of independent works done by
+// Lars Nyland and Andy Adinets
+
+///@todo: add cub's histogram as another option
+
+namespace MLCommon {
+namespace Stats {
+
+/** Default mapper which just returns the value of the data itself */
+template <typename DataT, typename IdxT>
+struct IdentityBinner {
+  DI int operator()(DataT val, IdxT row, IdxT col) { return int(val); }
+};
+
+/** Types of support histogram implementations */
+enum HistType {
+  /** use only global atomics */
+  HistTypeGmem,
+  /** uses shared mem atomics to reduce global traffic */
+  HistTypeSmem,
+  /** shared mem atomics but with bins to be 2B int's */
+  HistTypeSmemBits16,
+  /** shared mem atomics but with bins to ba 1B int's */
+  HistTypeSmemBits8,
+  /** shared mem atomics but with bins to be 4b int's */
+  HistTypeSmemBits4,
+  /** shared mem atomics but with bins to be 2b int's */
+  HistTypeSmemBits2,
+  /** shared mem atomics but with bins to be 1b int's */
+  HistTypeSmemBits1,
+  /** builds a hashmap of active bins in shared mem */
+  HistTypeSmemHash,
+  /** decide at runtime the best algo for the given inputs */
+  HistTypeAuto
+};
+
+static const int ThreadsPerBlock = 256;
+
+template <typename IdxT, int VecLen>
+dim3 computeGridDim(IdxT nrows, IdxT ncols) {
+  const auto maxBlks = 2 * getMultiProcessorCount();  // assume occupancy of 2
```

Comment:
For just 256 threads per block, feel free to have more blocks, especially for small numbers of columns and bins.

This can also be done in a different PR.

Reply:
I have generalized this through the use of cudart's occupancy API.